### ISSUE TITLE
[API Proposal]: A pattern for AOT-friendly source-generated `JsonConverterFactory`-based converters.

### ISSUE THREAD

**Issue by teo-tsirpanis** [dotnet/runtime#73124]
### Background and motivation

Using custom JSON converters that derive from `JsonConverterFactory` such as `JsonStringEnumConverter` does not always play well with AOT technologies even if we use the source generator. Consider the following code:

``` csharp
Console.WriteLine(JsonSerializer.Deserialize<MyEnum>("\"A\"", MyJsonContext.Default.MyEnum));

[JsonSerializable(typeof(MyEnum))]
internal partial class MyJsonContext : JsonSerializerContext { }

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MyEnum { A, B }
```

When compiled under NativeAOT, it will warn that `Using member 'System.Text.Json.Serialization.JsonStringEnumConverter.JsonStringEnumConverter()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. JSON serialization and deserialization might require types that cannot be statically analyzed and might need runtime code generation. Use System.Text.Json source generation for native AOT applications.`.

When ran under NativeAOT it will throw a `System.Reflection.MissingMetadataException: 'System.Text.Json.Serialization.Converters.EnumConverter<MyEnum>' is missing metadata.`. Fixing it needs fiddling with `rd.xml` files.

The reason for these failures is that there is not a continuous sequence of generic method calls from the source generator up until the creation of the `EnumConverter`, making the NativeAOT compiler's life difficult. The source-generated code calls `new JsonStringEnumConverter().CreateConverter(typeof(MyEnum), …)`, which calls `EnumConverterFactory.Create(typeof(MyEnum), …)`, [which uses the infamous `Type.MakeGenericType` to instantiate `EnumConverter<MyEnum>`](https://github.com/dotnet/runtime/blob/ec5e8b5fac0ca3524c01b826a048b0b293b63a7d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverterFactory.cs#L34).

We can avoid the dynamic type instantiation if we had a way to pass the type in the converter factory through a generic, which is what I am proposing.

### API Proposal

The JSON source generator will be taught so that if it sees a `[JsonConverterAttribute(typeof(TConverter))]` on a type `T` or a member of type `T` where `TConverter` has a static method with one generic parameter that accepts a `JsonSerializerOptions` and returns a `JsonConverter`, instead of emitting the equivalent of `new TConverter().CreateConverter(typeof(T), Options)`, it would emit `TConverter.CreateConverter<T>(Options)`.

Here's how `JsonStringEnumConverter` would be updated:

```csharp
namespace System.Text.Json;

public class JsonStringEnumConverter
{
    public static JsonConverter CreateConverter<T>(JsonSerializerOptions options) where T : struct, Enum;
}
```

### API Usage

The API is intended to be used by source-generated code as described above.

### Alternative Designs

* My first thought was to add an instance `JsonConverter CreateConverter<T>(JsonSerializerOptions options)` method to `JsonConverterFactory` but I realized mid-typing that `EnumConverter<T>` has generic constraints which cannot be bridged until dotnet/csharplang#6308 is implemented.
* If the scenario is not widely applicable, we could make a solution specific to enums by adding an overload to `JsonMetadataServices.GetEnumConverter` that allows specifying whether we want to handle enums as strings, and teaching the source generator to emit a call to this when `JsonStringEnumConverter` is applied.

**Comment by eiriktsarpalis** [comment]
You're correct to point out that `JsonConverterFactory` isn't particularly AOT-friendly when used in conjunction with generics, but the problem cannot be addressed simply by adding a generic factory method. As you mention there is the problem of generic constraints, but there's also other problems; for example, collection converters must know [both the type of the collection _and_ the type of the element](https://github.com/dotnet/runtime/blob/f1de614c74589fe8f682dda39889d189f2522504/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IAsyncEnumerableConverterFactory.cs#L25-L33). And this can grow in complexity to an arbitrary extent, e.g. consider [dictionary type converters](https://github.com/dotnet/runtime/blob/f1de614c74589fe8f682dda39889d189f2522504/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs#L193). This is a problem that cannot (at least in the case of .NET type systems) be overcome _without_ using reflection.

That being said, source generators very elegantly solve this problem, avoiding converter factories altogether and using [generic combinators acting on specific types](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.metadata.jsonmetadataservices.createdictionaryinfo?view=net-6.0). This leads to added boilerplate, of course, but we can ignore that since it's all handled by the source generator.

**Comment by jkotas** [comment]
> That being said, source generators very elegantly solve this problem, avoiding converter factories altogether and using [generic combinators acting on specific types](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.metadata.jsonmetadataservices.createdictionaryinfo?view=net-6.0).

I see we have `JsonMetadataServices.GetEnumConverter<T>`. How does one tell the source generator to use it for enum serialization?

Can the source generator check for `JsonStringEnumConverter` and replace it with AOT friendly call to `JsonMetadataServices.GetEnumConverter<T>` to make the example at the top just work?

**Comment by teo-tsirpanis** [comment]
`GetEnumConverter` will be used if we don't specify the custom converter attribute. The problem is that it does not support (de)serializing enums as strings. To support this we would need a new overload like `GetEnumConverter<T>(bool acceptStrings)`.

**Comment by jkotas** [comment]
Ok, I understand that we are out-of-time to address this for .NET 7. We should look into doing something about this soon.

The current experience is sub-par: The warning tells you to use Json source generation for AOT compatibility, but it is not actually possible to resolve the issue using Json source generation.

**Comment by eiriktsarpalis** [comment]
> > That being said, source generators very elegantly solve this problem, avoiding converter factories altogether and using [generic combinators acting on specific types](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.metadata.jsonmetadataservices.createdictionaryinfo?view=net-6.0).
> 
> I see we have `JsonMetadataServices.GetEnumConverter<T>`. How does one tell the source generator to use it for enum serialization?
> 
> Can the source generator check for `JsonStringEnumConverter` and replace it with AOT friendly call to `JsonMetadataServices.GetEnumConverter<T>` to make the example at the top just work?

We could do that, but fundamentally it is an ad-hoc solution that fixes the particular converter factory. Had I wanted to further customize enum serialization (e.g. by inheriting `JsonStringEnumConverter`) or, say, write a custom converter for generic arrays then it would still be impossible to make that work in AOT. `JsonConverterFactory` was not designed with such concerns in mind and it likely never will be able to address them in the general case. Which does makes me wonder whether there might be avenues to innovate and provide a replacement factory abstraction that is AOT friendly.

> The current experience is sub-par: The warning tells you to use Json source generation for AOT compatibility, but it is not actually possible to resolve the issue using Json source generation.

We should provide a PR that provides a better warning specifically for the `JsonStringEnumConverter` constructor. It's the only custom converter factory we're exposing so we don't need to change a lot.

**Comment by eiriktsarpalis** [comment]
I see a couple of components to be addressed in this issue:

1. An AOT-friendly pattern that supersedes `JsonConverterFactory`, this is obviously more long term.
2. Adding hardcoded support for `JsonStringEnumConverter`, which is pretty common in most serialization scenaria. Effectively we would have the source generator recognize & intercept `[JsonConverter(typeof(JsonStringEnumConverter))]` declarations and replace them with a call to a new `JsonMetadataServices.GetStringEnumConverter<T>` method which should at least serve to unblock that particular use case.

**Comment by eiriktsarpalis** [comment]
Leaving at 8.0.0 so that we at least address point number 2.

**Comment by kamyker** [comment]
> 'System.Text.Json.Serialization.Converters.EnumConverter<MyEnum>' is missing metadata.`. Fixing it needs fiddling with `rd.xml` files.
Since this issue is linked in the compiler: what exactly has to be done to fix it in .NET 7?

 I'm trying:
 ```
<Type Name="System.Text.Json.Serialization.Converters.EnumConverter`1[[MyAssembly.MyNamespace.MyType]]" Dynamic="Required All" />
 ```
 But that gives error:
 '''
Failed to load type 'MyAssembly.MyNamespace.MyType' from assembly 'System.Text.Json,
```

**Comment by eiriktsarpalis** [comment]
@kamyker I would recommend creating a separate issue, including a minimal reproducing app we could take a look at.

**Comment by kamyker** [comment]
> > 'System.Text.Json.Serialization.Converters.EnumConverter' is missing metadata.`. Fixing it needs fiddling with `rd.xml` files.
> 
> Since this issue is linked in the compiler: what exactly has to be done to fix it in .NET 7?
> 
> I'm trying:
> 
> ```
> <Assembly Name="System.Text.Json">
> <Type Name="System.Text.Json.Serialization.Converters.EnumConverter`1[[MyAssembly.MyNamespace.MyType]]" Dynamic="Required All" />
> </Assembly>
> ```
> 
> But that gives error:
> 
> ```
> Failed to load type 'MyAssembly.MyNamespace.MyType' from assembly 'System.Text.Json,
> ```

Nvm it has to be:
```
<Assembly Name="System.Text.Json">
<Type Name="System.Text.Json.Serialization.Converters.EnumConverter`1[[MyNamespace.MyType,MyAssembly]]" Dynamic="Required All" />
</Assembly>
```

**Comment by eiriktsarpalis** [comment]
One possible solution might be to simply replace `JsonConverterFactory` usage with public generic converters:
```C#
public class JsonStringEnumConverter<TEnum> : JsonConverter<TEnum>
    where TEnum : struct, Enum
{
}
```
So that type/property-level annotations effectively become linker/AOT-safe:
```C#
[JsonConverter(typeof(JsonStringEnumConverter<TEnum>)]
public enum MyEnum { A = 0, B = 1 }
```
Might be something to consider when designing the APIs for #63791.

**Comment by layomia** [comment]
We should design this feature with https://github.com/dotnet/runtime/issues/63791 in mind.

**Comment by teo-tsirpanis** [comment]
@Abdragiz I took a look at `GeoJsonConverterFactory` and it does not seem to be doing stuff that is unsupported in AOT, so it will likely work.

**Comment by BDisp** [comment]
I'm also having this issue with native AOT. Is there any solution?

**Comment by eiriktsarpalis** [comment]
@BDisp What we ended up doing in .NET 8 for `JsonStringEnumConverter` AOT support is exposing a [generic variant of the type](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonstringenumconverter-1?view=net-8.0). You could try emulating that approach.

See [this article](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8/runtime#source-generator) for more details.

**Comment by sliekens** [comment]
Is there a story for source-generated converters based on `JsonConverterFactory` when the type to convert is itself generic?

I have a wrapper type (`Extensible<TEnum>`) that wraps an enum and adds the ability to assign any string value.

``` csharp
Extensible<System.DayOfWeek> = "Doomsday";
```

The `Extensible<>` type is annotated with a `JsonConverterFactory`-based converter.

<details><summary>ExtensibleEnumJsonConverterFactory implementation details</summary>
<p>

``` csharp
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

internal sealed class ExtensibleEnumJsonConverterFactory : JsonConverterFactory
{
    public override bool CanConvert(Type typeToConvert)
    {
        if (typeToConvert == typeof(Extensible<System.DayOfWeek>))
        {
            return true;
        }

        return false;
    }

    public override JsonConverter? CreateConverter(
        Type typeToConvert,
        JsonSerializerOptions options
    )
    {
        if (typeToConvert == typeof(Extensible<System.DayOfWeek>))
        {
            return new ExtensibleEnumJsonConverter<System.DayOfWeek>();
        }

        return null;
    }
}

internal sealed class ExtensibleEnumJsonConverter<TEnum> : JsonConverter<Extensible<TEnum>>
    where TEnum : struct, Enum
{
    /// <inheritdoc />
    public override Extensible<TEnum> Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options
    )
    {
        string? name = reader.GetString();
        if (name is null)
        {
            ThrowHelper.ThrowInvalidOperationException("Expected a string but got null.");
        }

        return new Extensible<TEnum>(name);
    }

    /// <inheritdoc />
    public override void Write(
        Utf8JsonWriter writer,
        Extensible<TEnum> value,
        JsonSerializerOptions options
    )
    {
        writer.WriteStringValue(value.ToString());
    }

    /// <inheritdoc />
    public override Extensible<TEnum> ReadAsPropertyName(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options
    )
    {
        string? name = reader.GetString();
        if (name is null)
        {
            ThrowHelper.ThrowInvalidOperationException("Expected a string but got null.");
        }

        return new Extensible<TEnum>(name);
    }

    /// <inheritdoc />
    public override void WriteAsPropertyName(
        Utf8JsonWriter writer,
        Extensible<TEnum> value,
        JsonSerializerOptions options
    )
    {
        writer.WritePropertyName(value.ToString());
    }
}
```

</p>
</details> 



``` csharp
[JsonConverter(typeof(ExtensibleEnumJsonConverterFactory))]
public readonly struct Extensible<TEnum>(string name)
    where TEnum : struct, Enum
{
    //Not all code shown here

    public override readonly string ToString()
    {
        return name switch
        {
            not null => name,
            null when HasDefaultValue() => default(TEnum)!.ToString(),
            _ => ""
        };
    }
}
```

Now I am struggling to enable Native AOT with `JsonSerializerContext`.

``` csharp
[JsonSerializable(typeof(Extensible<DayOfWeek>), TypeInfoPropertyName = "ExtensibleDayOfWeek")]
public partial class SomeJsonContext : JsonSerializerContext
{
}
```

This fails with:

- **SYSLIB1220**: The 'JsonConverterAttribute' type 'ExtensibleEnumJsonConverterFactory' specified on member 'Extensible<System.DayOfWeek>' is not a converter type or does not contain an accessible parameterless constructor.
- **SYSLIB1030**: System.Text.Json source generator did not generate output for type

I looked around a bit and tried to understand the workaround that was implemented for regular enums. My (hasty) conclusion is that I can't do a similar workaround, because my type to convert is itself generic.
